cpp/src/arrow/util/cancel_test.cc

```diff
@@ -0,0 +1,202 @@
+    ASSERT_EQ(results, std::vector<int>({1, 2}));
+    token.RequestStop();
+    ASSERT_EQ(results, std::vector<int>({1, 2}));
+  }
+}
+
+TEST_F(CancelTest, StopCallbackMove) {
+  std::vector<int> results;
+  StopToken token;
+
+  StopCallback cb1(&token, [&](const Status& st) { results.push_back(1); });
+  const auto cb2 = std::move(cb1);
+
+  ASSERT_EQ(results.size(), 0);
+  token.RequestStop();
+  ASSERT_EQ(results, std::vector<int>{1});
+}
+
+TEST_F(CancelTest, ThreadedPollSuccess) {
+  constexpr int kNumThreads = 3;
+  constexpr int kMaxLoops = 30000;
+
+  std::vector<Status> results(kNumThreads);
+  std::vector<std::thread> threads;
+
+  StopToken token;
+  auto barrier = Future<void>::Make();
+
+  const auto worker_func = [&](int thread_num) {
+    ARROW_CHECK(barrier.Wait(kLongWait));
+    for (int i = 0; token.Poll().ok() && i < kMaxLoops; ++i) {
+    }
+    results[thread_num] = token.Poll();
+  };
+  for (int i = 0; i < kNumThreads; ++i) {
+    threads.emplace_back(std::bind(worker_func, i));
+  }
+
+  // Let the threads start
+  SleepFor(1e-3);
+  // Unblock all threads
+  barrier.MarkFinished();
+  for (auto& thread : threads) {
+    thread.join();
+  }
+
+  for (const auto& st : results) {
+    ASSERT_OK(st);
+  }
+}
+
+TEST_F(CancelTest, ThreadedPollCancel) {
+  constexpr int kNumThreads = 3;
+  constexpr int kMaxLoops = 1000000;
+
+  std::vector<Status> results(kNumThreads);
+  std::vector<std::thread> threads;
+
+  StopToken token;
+  auto barrier = Future<void>::Make();
+
+  const auto worker_func = [&](int thread_num) {
+    ARROW_CHECK(barrier.Wait(kLongWait));
+    for (int i = 0; token.Poll().ok() && i < kMaxLoops; ++i) {
+    }
+    results[thread_num] = token.Poll();
+  };
+
+  for (int i = 0; i < kNumThreads; ++i) {
+    threads.emplace_back(std::bind(worker_func, i));
+  }
+  // Let the threads start
+  SleepFor(1e-3);
+  // Unblock all threads
+  barrier.MarkFinished();
+  token.RequestStop(Status::IOError("Operation cancelled"));
+  for (auto& thread : threads) {
+    thread.join();
+  }
+
+  for (const auto& st : results) {
+    ASSERT_RAISES(IOError, st);
+  }
+}
+
+TEST_F(CancelTest, ThreadedSetCallbackCancel) {
+  constexpr int kIterations = 100;
+  constexpr double kMaxWait = 1e-3;
+
+  std::default_random_engine gen(42);
+  std::uniform_real_distribution<double> wait_dist(0.0, kMaxWait);
+
+  for (int i = 0; i < kIterations; ++i) {
+    Status result;
+
+    StopToken token;
+    auto barrier = Future<void>::Make();
+
+    const auto worker_func = [&]() {
+      ARROW_CHECK(barrier.Wait(kLongWait));
+      token.RequestStop(Status::IOError("Operation cancelled"));
+    };
+    std::thread thread(worker_func);
+
+    // Unblock thread
+    barrier.MarkFinished();
+    // Use a variable wait time to maximize potential synchronization issues
+    const auto wait_time = wait_dist(gen);
+    if (wait_time > kMaxWait * 0.5) {
+      SleepFor(wait_time);
```

Comment:
Instead of sleeping, just use a second barrier that is MarkFinished by the cancelling thread after the RequestStop call.

Reply:
Hmm, no, the purpose is to instantiate the `StopCallback` at any possible time, including while the other thread is running `token.RequestStop(...)`.